### ISSUE TITLE
Protocol whitelist in ytdl_hook

### ISSUE THREAD

**Issue by atalax** [mpv-player/mpv#5456]
The recent commits e6e6b0dcc7e9b0dbf35154a179b3dc1fcfcaff43, f8263e82cc74a9ac6530508bec39c7b0dc02568f, ce42a965330dfeb7d2f6c69ea42d35454105c828 and 2a0f9fc1588d8bda51a6340197c54285f3f62755 fix and issue whereby mpv could be convinced to play a "non-safe" URL from a remote source.

### Reproduction steps

An attacker convinces has the victim play an HTTP(S) URL.

```
$ mpv https://example.org/play.flv
```

The URL gets processed by the [ytdl_hook](https://github.com/mpv-player/mpv/blob/master/player/lua/ytdl_hook.lua) script.

youtube-dl attempts to extract videos from the URL by contacting the HTTP server, which
responds with something like (text/html mime typed) :

```
<html>
  <head>
  </head>
  <body>
    <video>
      <source src="av://lavfi:ladspa=file=/home/user/Downloads/libevil.so"></source>
    </video>
  </body>
</html>
```

As youtube-dl does not perform any validation on the extracted URLs for `<video>` tags, the `av://lavfi` URL gets passed back to the hook script.

_Note that there are likely [many ways](https://github.com/rg3/youtube-dl/tree/master/youtube_dl/extractor) in which youtube-dl can return "bad" URLs._

The hook script then passes the extracted URL to mpv, which does not apply the usual safe-protocol only [checks](https://github.com/mpv-player/mpv/blob/master/stream/stream.c#L219).

As shown in the example above, this URL can be, for instance, used to [dlopen()](https://github.com/FFmpeg/FFmpeg/blob/59365e634552e18bb7f844ba60cba2d62852fe88/libavfilter/af_ladspa.c#L412) arbitrary files on the filesystem using the ffmpeg lavfi ladspa plugin.

**Comment by atalax** [comment]
CVE-2018-6360 has been assigned to this issue

**Comment by jcowgill** [comment]
@atalax 2a0f9fc1588d seems to depend on 7eb342757320 (which would therefore needs adding to your list), but I am wondering if either of these commits are relevant to the security bug at all?

**Comment by CounterPillow** [comment]
I don't think 2a0f9fc is actually relevant to this CVE at all. The "whitelist" it mentions appears to be related to selecting a media stream from a manifest, and has no security whitelisting implications whatsoever.

**Comment by atalax** [comment]
Hmm, 2a0f9fc1588d8bda51a6340197c54285f3f62755 really doesn't seem relevant. Ping @wiiaboo

**Comment by wiiaboo** [comment]
It's a bugfix to https://github.com/mpv-player/mpv/commit/7eb34275732071acf779a0f241642411f91f78b7, yes, so not related to this issue.

**Comment by wiiaboo** [comment]
Correct string of commits needed is 828bd2963cd10a851e0a977809687aed4d377dc3 (demuxer-lavf-list) + e6e6b0dcc7e9b0dbf35154a179b3dc1fcfcaff43 (has merge conflicts with v0.27.0 due to a few changes since) + f8263e82cc74a9ac6530508bec39c7b0dc02568f + ce42a965330dfeb7d2f6c69ea42d35454105c828.

**Comment by kevmitch** [comment]
I'll let @wiiaboo give the final word, but it looks like that commit just adds a property. It provides information about which demuxers are available. It is not used in the subsequent white-listing commits which fix the actual problem.

**Comment by wiiaboo** [comment]
Yeah, nevermind. Being able to use the native dash demuxer is not necessary for the security fixes.

**Comment by jcowgill** [comment]
FYI I screwed up the fix I applied to Debian's 0.27 and 0.23 and broke YouTube playlists so don't take the patch from there just yet. Hopefully it will be fixed within a day or so.
